src/library/scanner/libraryscanner.cpp

```diff
@@ -452,6 +482,66 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
         emit(trackAdded(pTrack));
         emit(progressLoading(pTrack->getLocation()));
     } else {
-        qWarning() << "Track ("+pTrack->getLocation()+") could not be added";
+        qWarning()
+                << "Track (" + pTrack->getLocation() + ") could not be added";
+    }
+}
+
+bool LibraryScanner::changeScannerState(ScannerState newState) {
+    // Allowed State transitions:
+    // IDLE -> STARTING
+    // STARTING -> IDLE
+    // STARTING -> SCANNING
+    // SCANNING -> FINISHED
+    // FINISHED -> IDLE
+    // every state can change to CANCELING
+    // CANCELING -> IDLE
+    switch (newState) {
+    case IDLE:
+        // we are leaving STARTING  or CANCELING state
+        // m_state is already IDLE if a scan was canceled
+        m_state = IDLE;
+        m_stateMutex.unlock();
+        return true;
+    case STARTING:
+        // we need to lock the mutex during the STARTING state
+        // to prevent loosing cancel commands or start the scanner
+        // twice
+        if (m_stateMutex.tryLock()) {
+            if (m_state != IDLE) {
+                qDebug() << "LibraryScanner: Scan already in progress.";
+                m_stateMutex.unlock();
+                return false;
+            }
+            m_state = STARTING;
+            return true;
+        } else {
+            qDebug() << "LibraryScanner: mutex locked, state =" << m_state;
+            return false;
+        }
+    case SCANNING:
+        DEBUG_ASSERT(m_state == STARTING);
```

Comment:
> Look at this like unit-tests

runtime checks are not at all like unit tests.  I understand how this could catch regressions, but it's the least-desirable way to test for regressions.